sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java

```diff
@@ -584,7 +584,7 @@ private int getCreditsToAdd() {
         if (subscriber == null || request == 0) {
             credits = 0;
         } else if (request == Long.MAX_VALUE) {
-            credits = 1;
+            credits = prefetch;
```

Comment:
I guess the value of `prefetch` is the value of `preFetch` in your demo code?

Reply:
that's correct.

Reply:
The JavaDoc for this method should be updated to reflect this behavior. Apparently, this behavior of adding just 1 credit at a time is similar to what Track 1 did. There may have been a reason for doing it this way that @conniey may be aware of.

Reply:
Also, have we measured the perf improvement with this change and has this made the client less chatty?

Reply:
Another question is, what's the theoretical max number of messages in the cache when we add credits in hundreds.
For instance, if the cache already has a lot of messages and we add a few hundred credits after each message is consumed?

Reply:
> srnagar 
Also, have we measured the perf improvement with this change and has this made the client less chatty?

@srnagar, yep, I've used the `receiveevents` test in the perf framework which exercises the no-backpressure code path. The performance of T2 in this code-path improved drastically and T1 and T2 are very close now. 

Below is the result of perf testing this specific code path with a prefetch of 100 and a count of 1800 (the test is the same as `receiveevents` perf test, I slightly modified it to be more debuggable and print extra info).

[Executed for 5 times]
 
```
TRACK_1: Start--
Took 14 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 9 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 5 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 5 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 5 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 12 seconds to receive 1800 events
TRACK_2: End--

TRACK_1: Start--
Took 8 seconds to receive 1800 events
TRACK_1: End--

TRACK_2: Start--
Took 7 seconds to receive 1800 events
TRACK_2: End--
```

Reply:
> srnagar .... Apparently, this behavior of adding just 1 credit at a time is similar to what Track 1 did. There may have been a reason for doing it this way that @conniey may be aware of.

Connie and I visited that track-1 code, track-1 has a special code path that adds 1 credit at a time, but that only accumulates the credits locally. Once the accumulated credits hit a cap, the actual Flow will be sent. So, track2 won't send one flow for each credit.

Reply:
> YijunXieMS 
Another question is, what's the theoretical max number of messages in the cache when we add credits in hundreds.
For instance, if the cache already has a lot of messages and we add a few hundred credits after each message is consumed?

`what's the theoretical max number of messages in the cache` - It depends on the operators that the user applied in the chain in their application side, different operators can have their own queue, holding messages to be delivered for their immediate subscriber. Our processor also has its own queue (say q1), when we receive a "request" for messages from the downstream, the credit will be computed using the size of q1 and "request"ed amount, that's the back-pressure aware flow, MAX_VALUE is the no-back-pressure flow.